src/java/org/apache/cassandra/net/Message.java

```diff
@@ -604,7 +604,25 @@ public Message<T> build()
             if (payload == null)
                 throw new IllegalArgumentException();
 
-            return new Message<>(new Header(hasId ? id : nextId(), verb, from, createdAtNanos, expiresAtNanos, flags, params), payload);
+            return doBuild(hasId ? id : nextId());
+        }
+
+        public int currentSize(int version)
```

Comment:
So basically we build the message twice now - once we calculate the size, and another when want to build the actual response, is this ok performance wise? This was one of the reasons to use register to the outbound sink (tradeoff being complexity of the overall approach, just making sure you think the redundant message build is worth it)

Reply:
Point taken. The most expensive computation is probably the payload size, what if we cache it?

Reply:
Not sure how caching the payload size would help. It seems you call it only once? I was also thinking building the message to calculate the size has a memory hit. 

The size calculation of the message and including it in the headers is a chicken and egg problem. tbh don't see an effective way around it.

Reply:
> It seems you call it only once?

Once per message, but we build the message twice. The reason we cache it in the message is exactly the fact that it's the most expensive size computation.

> I was also thinking building the message to calculate the size has a memory hit.

Building the `Message` is just a bunch of objects assignment, isn't it?

Reply:
> Once per message, but we build the message twice. The reason we cache it in the message is exactly the fact that it's the most expensive size computation.

Exactly - that's why I wasn't sure how caching would help, maybe I misunderstood your suggestion. 

> Building the Message is just a bunch of objects assignment, isn't it?

Fair point. No more concerns here.